File: src/divmod_tutorial.py
"""Complete guide to divmod and related division operations in Python."""


# ============================================================================
# 1. DIVMOD BASICS
# ============================================================================

def divmod_basics():
    """divmod(a, b) returns (quotient, remainder) as a tuple."""
    
    # divmod(17, 5) means: 17 ÷ 5 = 3 remainder 2
    quotient, remainder = divmod(17, 5)
    print(f"17 ÷ 5 = {quotient} remainder {remainder}")  # 3 remainder 2
    
    # Equivalent to:
    q = 17 // 5  # Floor division (quotient)
    r = 17 % 5   # Modulo (remainder)
    print(f"Using // and %: {q} remainder {r}")
    
    # Why use divmod? It's more efficient (one operation vs two)
    result = divmod(100, 7)
    print(f"divmod(100, 7) = {result}")  # (14, 2)


# ============================================================================
# 2. PRACTICAL USE CASES
# ============================================================================

def convert_seconds_to_time(total_seconds):
    """Convert seconds to hours, minutes, seconds."""
    hours, remainder = divmod(total_seconds, 3600)
    minutes, seconds = divmod(remainder, 60)
    return f"{hours}h {minutes}m {seconds}s"


def convert_cents_to_dollars(cents):
    """Convert cents to dollars and cents."""
    dollars, remaining_cents = divmod(cents, 100)
    return f"${dollars}.{remaining_cents:02d}"


def split_into_groups(total_items, group_size):
    """Calculate full groups and leftover items."""
    full_groups, leftover = divmod(total_items, group_size)
    return f"{full_groups} groups of {group_size}, {leftover} left over"


# ============================================================================
# 3. DIVISION OPERATORS COMPARISON
# ============================================================================

def division_operators():
    """Compare all division operators in Python."""
    a, b = 17, 5
    
    print(f"a = {a}, b = {b}\n")
    
    # Regular division (float result)
    print(f"a / b  = {a / b}")           # 3.4 (true division)
    
    # Floor division (rounds down to nearest integer)
    print(f"a // b = {a // b}")          # 3 (quotient)
    
    # Modulo (remainder)
    print(f"a % b  = {a % b}")           # 2 (remainder)
    
    # divmod (both at once)
    print(f"divmod(a, b) = {divmod(a, b)}")  # (3, 2)
    
    # Verify: quotient * divisor + remainder = dividend
    q, r = divmod(a, b)
    print(f"\nVerify: {q} * {b} + {r} = {q * b + r}")  # 3 * 5 + 2 = 17


# ============================================================================
# 4. NEGATIVE NUMBERS
# ============================================================================

def negative_numbers():
    """divmod with negative numbers (floors toward negative infinity)."""
    
    print("Positive numbers:")
    print(f"divmod(17, 5) = {divmod(17, 5)}")    # (3, 2)
    
    print("\nNegative dividend:")
    print(f"divmod(-17, 5) = {divmod(-17, 5)}")  # (-4, 3)
    # -17 = -4 * 5 + 3 = -20 + 3 = -17 ✓
    
    print("\nNegative divisor:")
    print(f"divmod(17, -5) = {divmod(17, -5)}")  # (-4, -3)
    # 17 = -4 * -5 + -3 = 20 - 3 = 17 ✓
    
    print("\nBoth negative:")
    print(f"divmod(-17, -5) = {divmod(-17, -5)}")  # (3, -2)
    # -17 = 3 * -5 + -2 = -15 - 2 = -17 ✓


# ============================================================================
# 5. REAL-WORLD EXAMPLES
# ============================================================================

def format_bytes(bytes_count):
    """Convert bytes to KB, MB, GB."""
    gb, remainder = divmod(bytes_count, 1024**3)
    mb, remainder = divmod(remainder, 1024**2)
    kb, b = divmod(remainder, 1024)
    
    parts = []
    if gb: parts.append(f"{gb}GB")
    if mb: parts.append(f"{mb}MB")
    if kb: parts.append(f"{kb}KB")
    if b: parts.append(f"{b}B")
    
    return " ".join(parts) or "0B"


def paginate(total_items, items_per_page):
    """Calculate pagination info."""
    total_pages, extra = divmod(total_items, items_per_page)
    if extra > 0:
        total_pages += 1  # Need one more page for remaining items
    return total_pages


def decimal_to_base(number, base):
    """Convert decimal number to any base (2-36)."""
    if number == 0:
        return "0"
    
    digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    result = []
    
    while number > 0:
        number, remainder = divmod(number, base)
        result.append(digits[remainder])
    
    return "".join(reversed(result))


# ============================================================================
# 6. PERFORMANCE COMPARISON
# ============================================================================

def performance_test():
    """divmod is faster than separate // and % operations."""
    import timeit
    
    # Using divmod
    time_divmod = timeit.timeit(
        "divmod(123456789, 7)",
        number=1_000_000
    )
    
    # Using separate operations
    time_separate = timeit.timeit(
        "123456789 // 7; 123456789 % 7",
        number=1_000_000
    )
    
    print(f"divmod: {time_divmod:.4f}s")
    print(f"// and %: {time_separate:.4f}s")
    print(f"divmod is {time_separate/time_divmod:.2f}x faster")


# ============================================================================
# DEMO
# ============================================================================

if __name__ == "__main__":
    print("=" * 60)
    print("1. DIVMOD BASICS")
    print("=" * 60)
    divmod_basics()
    
    print("\n" + "=" * 60)
    print("2. PRACTICAL EXAMPLES")
    print("=" * 60)
    print(convert_seconds_to_time(3665))
    print(convert_cents_to_dollars(1234))
    print(split_into_groups(47, 5))
    
    print("\n" + "=" * 60)
    print("3. DIVISION OPERATORS")
    print("=" * 60)
    division_operators()
    
    print("\n" + "=" * 60)
    print("4. NEGATIVE NUMBERS")
    print("=" * 60)
    negative_numbers()
    
    print("\n" + "=" * 60)
    print("5. REAL-WORLD EXAMPLES")
    print("=" * 60)
    print(f"Bytes: {format_bytes(5_368_709_120)}")
    print(f"Pages needed for 100 items (10/page): {paginate(100, 10)}")
    print(f"255 in binary: {decimal_to_base(255, 2)}")
    print(f"255 in hex: {decimal_to_base(255, 16)}")
    
    print("\n" + "=" * 60)
    print("6. PERFORMANCE")
    print("=" * 60)
    performance_test()

File: src/DivModTutorial.java
/**
 * Complete guide to division operations in Java (Python divmod equivalent)
 * 
 * Java doesn't have a built-in divmod function, but you can:
 * 1. Use / and % operators separately
 * 2. Create a custom divmod method
 * 3. Use Math.floorDiv() and Math.floorMod() for floor division
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class DivModTutorial {

    // ========================================================================
    // 1. CUSTOM DIVMOD CLASS
    // ========================================================================
    
    static class DivModResult {
        final int quotient;
        final int remainder;
        
        DivModResult(int quotient, int remainder) {
            this.quotient = quotient;
            this.remainder = remainder;
        }
        
        @Override
        public String toString() {
            return "(" + quotient + ", " + remainder + ")";
        }
    }
    
    static DivModResult divmod(int dividend, int divisor) {
        int quotient = dividend / divisor;
        int remainder = dividend % divisor;
        return new DivModResult(quotient, remainder);
    }
    
    static class DivModResultLong {
        final long quotient;
        final long remainder;
        
        DivModResultLong(long quotient, long remainder) {
            this.quotient = quotient;
            this.remainder = remainder;
        }
        
        @Override
        public String toString() {
            return "(" + quotient + ", " + remainder + ")";
        }
    }
    
    static DivModResultLong divmod(long dividend, long divisor) {
        long quotient = dividend / divisor;
        long remainder = dividend % divisor;
        return new DivModResultLong(quotient, remainder);
    }

    // ========================================================================
    // 2. DIVISION OPERATORS IN JAVA
    // ========================================================================
    
    static void divisionOperators() {
        System.out.println("JAVA DIVISION OPERATORS");
        System.out.println("=======================\n");
        
        int a = 17, b = 5;
        System.out.println("a = " + a + ", b = " + b + "\n");
        
        // Regular division (truncates toward zero for integers)
        System.out.println("a / b  = " + (a / b));           // 3 (integer division)
        System.out.println("(double)a / b = " + ((double)a / b)); // 3.4 (float division)
        
        // Modulo (remainder)
        System.out.println("a % b  = " + (a % b));           // 2
        
        // Custom divmod
        DivModResult result = divmod(a, b);
        System.out.println("divmod(a, b) = " + result);      // (3, 2)
        
        // Verify
        System.out.println("\nVerify: " + result.quotient + " * " + b + " + " + 
                          result.remainder + " = " + (result.quotient * b + result.remainder));
    }

    // ========================================================================
    // 3. PRACTICAL USE CASES
    // ========================================================================
    
    static String convertSecondsToTime(int totalSeconds) {
        DivModResult hm = divmod(totalSeconds, 3600);
        int hours = hm.quotient;
        
        DivModResult ms = divmod(hm.remainder, 60);
        int minutes = ms.quotient;
        int seconds = ms.remainder;
        
        return hours + "h " + minutes + "m " + seconds + "s";
    }
    
    static String convertCentsToDollars(int cents) {
        DivModResult result = divmod(cents, 100);
        return String.format("$%d.%02d", result.quotient, result.remainder);
    }
    
    static String splitIntoGroups(int totalItems, int groupSize) {
        DivModResult result = divmod(totalItems, groupSize);
        return result.quotient + " groups of " + groupSize + ", " + 
               result.remainder + " left over";
    }

    // ========================================================================
    // 4. NEGATIVE NUMBERS - KEY DIFFERENCE!
    // ========================================================================
    
    static void negativeNumbers() {
        System.out.println("NEGATIVE NUMBERS - JAVA vs PYTHON");
        System.out.println("===================================\n");
        
        System.out.println("Java / and % truncate toward ZERO");
        System.out.println("Python // and % floor toward NEGATIVE INFINITY\n");
        
        // Java behavior
        System.out.println("JAVA:");
        System.out.println("17 / 5   = " + (17 / 5));      // 3
        System.out.println("-17 / 5  = " + (-17 / 5));     // -3 (truncates toward zero)
        System.out.println("17 % 5   = " + (17 % 5));      // 2
        System.out.println("-17 % 5  = " + (-17 % 5));     // -2 (sign matches dividend)
        
        // Python-like behavior using Math.floorDiv and Math.floorMod
        System.out.println("\nJAVA (Python-like with Math.floorDiv/floorMod):");
        System.out.println("Math.floorDiv(17, 5)  = " + Math.floorDiv(17, 5));    // 3
        System.out.println("Math.floorDiv(-17, 5) = " + Math.floorDiv(-17, 5));   // -4 (floors down)
        System.out.println("Math.floorMod(17, 5)  = " + Math.floorMod(17, 5));    // 2
        System.out.println("Math.floorMod(-17, 5) = " + Math.floorMod(-17, 5));   // 3 (always positive)
    }

    // ========================================================================
    // 5. REAL-WORLD EXAMPLES
    // ========================================================================
    
    static String formatBytes(long bytesCount) {
        long gb = 1024L * 1024 * 1024;
        long mb = 1024L * 1024;
        long kb = 1024L;
        
        DivModResultLong r1 = divmod(bytesCount, gb);
        DivModResultLong r2 = divmod(r1.remainder, mb);
        DivModResultLong r3 = divmod(r2.remainder, kb);
        
        List<String> parts = new ArrayList<>();
        if (r1.quotient > 0) parts.add(r1.quotient + "GB");
        if (r2.quotient > 0) parts.add(r2.quotient + "MB");
        if (r3.quotient > 0) parts.add(r3.quotient + "KB");
        if (r3.remainder > 0) parts.add(r3.remainder + "B");
        
        return parts.isEmpty() ? "0B" : String.join(" ", parts);
    }
    
    static int paginate(int totalItems, int itemsPerPage) {
        DivModResult result = divmod(totalItems, itemsPerPage);
        int totalPages = result.quotient;
        if (result.remainder > 0) {
            totalPages++;
        }
        return totalPages;
    }
    
    static String decimalToBase(int number, int base) {
        if (number == 0) return "0";
        
        String digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        List<Character> result = new ArrayList<>();
        
        while (number > 0) {
            DivModResult dm = divmod(number, base);
            number = dm.quotient;
            result.add(digits.charAt(dm.remainder));
        }
        
        Collections.reverse(result);
        StringBuilder sb = new StringBuilder();
        for (char c : result) sb.append(c);
        return sb.toString();
    }

    // ========================================================================
    // 6. PERFORMANCE COMPARISON
    // ========================================================================
    
    static void performanceTest() {
        int iterations = 10_000_000;
        int a = 123456789, b = 7;
        
        // Using separate operations
        long start1 = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            int q = a / b;
            int r = a % b;
        }
        long time1 = System.nanoTime() - start1;
        
        // Using divmod method
        long start2 = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            DivModResult result = divmod(a, b);
        }
        long time2 = System.nanoTime() - start2;
        
        System.out.println("Separate / and %: " + time1 / 1_000_000 + "ms");
        System.out.println("divmod method:    " + time2 / 1_000_000 + "ms");
        System.out.println("\nNote: In Java, separate operations are usually faster");
        System.out.println("because the JVM optimizes them. Use divmod for clarity.");
    }

    // ========================================================================
    // MAIN DEMO
    // ========================================================================
    
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("1. DIVISION OPERATORS");
        System.out.println("=".repeat(60));
        divisionOperators();
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("2. PRACTICAL EXAMPLES");
        System.out.println("=".repeat(60));
        System.out.println(convertSecondsToTime(3665));
        System.out.println(convertCentsToDollars(1234));
        System.out.println(splitIntoGroups(47, 5));
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("3. NEGATIVE NUMBERS");
        System.out.println("=".repeat(60));
        negativeNumbers();
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("4. REAL-WORLD EXAMPLES");
        System.out.println("=".repeat(60));
        System.out.println("Bytes: " + formatBytes(5_368_709_120L));
        System.out.println("Pages needed for 100 items (10/page): " + paginate(100, 10));
        System.out.println("255 in binary: " + decimalToBase(255, 2));
        System.out.println("255 in hex: " + decimalToBase(255, 16));
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("5. PERFORMANCE");
        System.out.println("=".repeat(60));
        performanceTest();
    }
}
